core/src/main/java/com/linecorp/armeria/internal/common/BodySubscriber.java

```diff
@@ -0,0 +1,306 @@
+            final Throwable cause = this.cause;
+            if (cause != null) {
+                onError0(cause, downstream);
+            } else if (completing) {
+                onComplete0(downstream);
+            }
+        } catch (Throwable t) {
+            throwIfFatal(t);
+            logger.warn("Subscriber should not throw an exception. subscriber: {}", downstream, t);
+        }
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (upstream != null) {
+            subscription.cancel();
+            return;
+        }
+        upstream = subscription;
+        if (cancelCalled) {
+            subscription.cancel();
+            return;
+        }
+        if (pendingRequests > 0) {
+            subscription.request(pendingRequests);
+        }
+    }
+
+    @Override
+    public void request(long n) {
+        if (n <= 0) {
+            upstreamMessage.abort(new IllegalArgumentException(
+                    "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+            return;
+        }
+        if (upstreamExecutor.inEventLoop()) {
+            request0(n);
+        } else {
+            upstreamExecutor.execute(() -> request0(n));
+        }
+    }
+
+    private void request0(long n) {
+        final Subscription upstream = this.upstream;
+        if (upstream == null) {
+            pendingRequests = LongMath.saturatedAdd(n, pendingRequests);
+        } else {
+            upstream.request(n);
+        }
+    }
+
+    @Override
+    public final void cancel() {
+        if (cancelCalled) {
+            return;
+        }
+        cancelCalled = true;
+        if (!notifyCancellation) {
+            downstream = NoopSubscriber.get();
+        }
+        completeTrailers(EMPTY_TRAILERS);
+        final Subscription upstream = this.upstream;
+        if (upstream != null) {
+            upstream.cancel();
+        }
+    }
+
+    @Override
+    public void onNext(HttpObject httpObject) {
+        if (httpObject instanceof HttpHeaders) {
+            final HttpHeaders trailers = (HttpHeaders) httpObject;
+            completeTrailers(trailers);
+            return;
+        }
+
+        final Subscriber<? super HttpData> downstream = this.downstream;
+        assert downstream != null;
+        assert httpObject instanceof HttpData;
+
+        final EventExecutor executor = this.executor;
+        if (executor.inEventLoop()) {
+            onNext0((HttpData) httpObject);
+        } else {
+            executor.execute(() -> onNext0((HttpData) httpObject));
+        }
+    }
+
+    private void onNext0(HttpData httpData) {
+        wroteAny = true;
+        if (!usePooledObject) {
+            httpData = PooledObjects.copyAndClose(httpData);
+        }
+        downstream.onNext(httpData);
+    }
+
+    @Override
+    public void onComplete() {
+        final EventExecutor executor = this.executor;
```

Comment:
I think we should call `maybeCompleteHeaders(null)` here as the original implementation did.